tracing-subscriber/src/registry/sharded.rs

```diff
@@ -0,0 +1,253 @@
+            extensions: RwLock::new(ExtensionsInner::new()),
+        };
+        let id = self.insert(s).expect("Unable to allocate another span");
+        idx_to_id(id)
+    }
+
+    /// This is intentionally not implemented, as recording fields
+    /// on a span is the responsibility of layers atop of this registry.
+    #[inline]
+    fn record(&self, _: &span::Id, _: &span::Record<'_>) {}
+
+    fn record_follows_from(&self, _span: &span::Id, _follows: &span::Id) {}
+
+    /// This is intentionally not implemented, as recording events
+    /// is the responsibility of layers atop of this registry.
+    fn event(&self, _: &Event<'_>) {}
+
+    fn enter(&self, id: &span::Id) {
+        CURRENT_SPANS.with(|spans| {
+            spans.borrow_mut().push(self.clone_span(id));
+        })
+    }
+
+    fn exit(&self, id: &span::Id) {
+        if let Some(id) = CURRENT_SPANS.with(|spans| spans.borrow_mut().pop(id)) {
+            dispatcher::get_default(|dispatch| dispatch.try_close(id.clone()));
+        }
+    }
+
+    fn clone_span(&self, id: &span::Id) -> span::Id {
+        let refs = self
+            .get(&id)
+            .unwrap_or_else(|| panic!("tried to clone {:?}, but no span exists with that ID", id))
+            .ref_count
+            .fetch_add(1, Ordering::Relaxed);
+        assert!(refs != 0, "tried to clone a span that already closed");
+        id.clone()
+    }
+
+    fn current_span(&self) -> Current {
+        CURRENT_SPANS
+            .with(|spans| {
+                let spans = spans.borrow();
+                let id = spans.current()?;
+                let span = self.get(id)?;
+                Some(Current::new(id.clone(), span.metadata))
+            })
+            .unwrap_or_else(Current::none)
+    }
+
+    /// Decrements the reference count of the span with the given `id`, and
+    /// removes the span if it is zero.
+    ///
+    /// The allocated span slot will be reused when a new span is created.
+    fn try_close(&self, id: span::Id) -> bool {
+        let span = match self.get(&id) {
+            Some(span) => span,
+            None if std::thread::panicking() => return false,
+            None => panic!("tried to drop a ref to {:?}, but no such span exists!", id),
+        };
+
+        let refs = span.ref_count.fetch_sub(1, Ordering::Release);
+        if !std::thread::panicking() {
+            assert!(refs < std::usize::MAX, "reference count overflow!");
+        }
+        if refs > 1 {
+            return false;
+        }
+
+        // Synchronize only if we are actually removing the span (stolen
+        // from std::Arc);
+        fence(Ordering::Acquire);
+        self.spans.remove(id_to_idx(&id));
+        true
+    }
+}
+
+impl<'a> LookupSpan<'a> for Registry {
+    type Data = Data<'a>;
+
+    fn span_data(&'a self, id: &Id) -> Option<Self::Data> {
+        let inner = self.get(id)?;
+        Some(Data { inner })
+    }
+}
+
+// === impl DataInner ===
+
+impl Drop for DataInner {
```

Comment:
I don't understand what purpose this `Drop` impl serves.

Reply:
@jonhoo I added an implementation comment explaining this in d588772373ed4042eb8c6f3ba60907cde461951f